src/aarch64/ffi.c

```diff
@@ -699,10 +702,29 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
   struct call_frame *frame;
   void *stack, *rvalue;
   struct arg_state state;
-  size_t stack_bytes, rtype_size, rsize;
+  size_t stack_bytes, rtype_size, rsize, total_length;
+  size_t context_length, context_offset, frame_length, frame_offset;
+  size_t rvalue_length, rvalue_offset, stack_length, stack_offset;
   int i, nargs, flags, isvariadic = 0;
+  char *cursor;
   ffi_type *rtype;
 
+#ifdef __CHERI_PURE_CAPABILITY__
+#define SUBOBJECT_ALIGNMENT(length)					\
+	~__builtin_cheri_representable_alignment_mask(length) + 1
+#define SUBOBJECT_LENGTH(size)						\
+	__builtin_cheri_round_representable_length(size)
+#else
+#define SUBOBJECT_ALIGNMENT 1
```

Comment:
I think it needs to take an argument here, otherwise it won't pre-process?
```suggestion
#define SUBOBJECT_ALIGNMENT(length) 1
```

Reply:
Oops, yes. @kwitaszczyk while you're changing this, could you rename SUBOBJECT_LENGTH's parameter to be consistent between the two? I wasn't going to nitpick it as the only change, but now it's not :)

Reply:
> I think it needs to take an argument here, otherwise it won't pre-process?

Yes, I didn't run an aarch64 build. Thanks!

Reply:
> Oops, yes. @kwitaszczyk while you're changing this, could you rename SUBOBJECT_LENGTH's parameter to be consistent between the two? I wasn't going to nitpick it as the only change, but now it's not :)

Just to make sure: do you mean that `SUBOBJECT_LENGTH()` and `SUBOBJECT_ALIGNMENT()` should use the same argument name, or they should use different argument name but the same across ABIs they're compiled for (regardless of `__CHERI_PURE_CAPABILITY__`)? It was broken in both cases. :) I'd be leaning to keep the latter as `SUBOBJECT_LAYOUT()` passes a requested size to `SUBOBJECT_LENGTH()` but it passes a representable length to `SUBOBJECT_ALIGNMENT()`.

Reply:
Specifically the latter, though not doing the former as well does look a bit odd, even if you want to make a distinction between size and length (which I don't think exists, because a size is a length, even if not representable).

Reply:
Sounds good. It's fixed now.